### PR TITLE
Policy: Deny VM Unmanaged Disks

### PR DESCRIPTION
**Pull request by Springstone**
## Overview/Summary

New policy and assignment to deny the deployment of virtual machines using Unamanaged Disks.

Addressed story: [AB#25733](https://dev.azure.com/CSUSolEng/2a9d5a5a-8998-4ad0-81c8-ef3045e4da97/_workitems/edit/25733)
Tested successfully.

## This PR fixes/adds/changes/removes

1. New policy and assignment : **Deny-VM-with-unmanaged-disk**

### Breaking Changes

1. N/A

#### Azure Public

[Deploy To Azure](https://portal.azure.com/#view/Microsoft_Azure_CreateUIDef/CustomDeploymentBlade/uri/https%3A%2F%2Fraw.githubusercontent.com%2FSpringstone%2FEnterprise-Scale%2Fpolicy-refresh-25733%2FeslzArm%2FeslzArm.json/uiFormDefinitionUri/https%3A%2F%2Fraw.githubusercontent.com%2FSpringstone%2FEnterprise-Scale%2Fpolicy-refresh-25733%2FeslzArm%2Feslz-portal.json)

## As part of this Pull Request I have

- [X] Checked for duplicate [Pull Requests](https://github.com/Azure/Enterprise-Scale/pulls)
- [ ] Associated it with relevant [issues](https://github.com/Azure/Enterprise-Scale/issues), for tracking and closure.
- [X] Ensured my code/branch is up-to-date with the latest changes in the `main` [branch](https://github.com/Azure/Enterprise-Scale/tree/main)
- [X] Performed testing and provided evidence.
- [X] Ensured [contribution guidance](https://github.com/Azure/Enterprise-Scale/wiki/ALZ-Contribution-Guide) is followed.
- [X] Updated relevant and associated documentation.
- [X] Updated the ["What's New?"](https://github.com/Azure/Enterprise-Scale/wiki/Whats-new) wiki page (located: `/docs/wiki/whats-new.md`)

### CODE REVIEW

File: src/resources/Microsoft.Authorization/policyDefinitions/Deny-VMUnmanagedDisk.json [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+                                "field": "type",
+                                "equals": "Microsoft.Compute/virtualMachineScaleSets"
+                            },
+                            {
+                                "field": "Microsoft.Compute/virtualMachineScaleSets/virtualMachineProfile.storageProfile.osDisk.managedDisk",
+                                "exists": "false"
+                            }
+                        ]
+                    }
+                ]
```
Comment (jtracey93) [inline src/resources/Microsoft.Authorization/policyDefinitions/Deny-VMUnmanagedDisk.json:60]: @aarunraaj @Springstone what about data disks? This just covers OS disks?

File: src/resources/Microsoft.Authorization/policyDefinitions/Deny-VMUnmanagedDisk.json [hunk 2]
```diff
@@ -0,0 +1,67 @@
+{
+    "name": "Deny-VM-with-unmanaged-disk",
+    "apiVersion": "2021-06-01",
+    "scope": "null",
+    "properties": {
+        "policyType": "Custom",
+        "mode": "all",
+        "displayName": "Deny virtual machines and virtual machine scalesets not using managed disk",
```
Comment (jtracey93) [inline src/resources/Microsoft.Authorization/policyDefinitions/Deny-VMUnmanagedDisk.json:8]:
```suggestion
        "displayName": "Deny Virtual Machines and Virtual Machine Scale Sets from not using Managed Disks",
```

File: src/resources/Microsoft.Authorization/policyDefinitions/Deny-VMUnmanagedDisk.json [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        "description": "Deny virtual machines not using managed disk. It checks the managedDisk property on virtual machine OS, and Datadisk fields.",
```
Comment (jtracey93) [inline src/resources/Microsoft.Authorization/policyDefinitions/Deny-VMUnmanagedDisk.json:9]:
```suggestion
        "description": "Deny Virtual Machines and Virtual Machine Scale Sets from not using Managed Disks. It checks the managedDisk property on Virtual Machine & Virtual Machine Scale Sets OS, and Data Disk fields.",
```

This also isnt true about data disks in its current form

File: src/resources/Microsoft.Authorization/policyDefinitions/Deny-VMUnmanagedDisk.json [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        "metadata": {
+            "version": "1.0.0.0",
```
Comment (jtracey93) [inline src/resources/Microsoft.Authorization/policyDefinitions/Deny-VMUnmanagedDisk.json:11]:
```suggestion
            "version": "1.0.0",
```

File: src/resources/Microsoft.Authorization/policyDefinitions/Deny-VMUnmanagedDisk.json [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+            "category": "Compute",
+            "source": "https://github.com/Azure/Enterprise-Scale/",
+            "alzCloudEnvironments": [
+                "AzureCloud"
+            ]
```
Comment (jtracey93) [inline src/resources/Microsoft.Authorization/policyDefinitions/Deny-VMUnmanagedDisk.json:16]: Should the other clouds be supported here as we know managed and unmanaged disk are available in china and us gov?

File: src/resources/Microsoft.Authorization/policyDefinitions/Deny-VMUnmanagedDisk.json [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+                            {
+                                "field": "Microsoft.Compute/virtualMachineScaleSets/virtualMachineProfile.storageProfile.osDisk.managedDisk",
+                                "exists": "false"
+                            }
+                        ]
+                    }
+                ]
+            },
+            "then": {
+                "effect": "[parameters('effectType')]"
```
Comment (jtracey93) [inline src/resources/Microsoft.Authorization/policyDefinitions/Deny-VMUnmanagedDisk.json:63]:
```suggestion
                "effect": "[[parameters('effectType')]"
```

File: docs/wiki/ALZ-Policies.md [hunk 7]
```diff
@@ -70,6 +70,7 @@ The table below provides the specific **Custom** and **Built-in** **policy defin
...
 | **Deny the Deployment of Classic Resources** | **Not allowed resource types** | `Policy Definition`, **Built-in**     | Denies deployment of classic resource types under the assigned scope                                                                                                                                                                                                                              | Deny                   | 2.0.0   |
+| **Deny virtual machines and virtual machine scale sets not using managed disk** | **Deny virtual machines and virtual machine scale sets not using managed disk** | `Policy Definition`, **Custom**     | Deny virtual machines not using managed disk. It checks the managedDisk property on virtual machine OS, and Datadisk fields.                                                                                                                                                                                                                         | Deny                   | 1.0.0   |
```
Comment (jtracey93) [inline docs/wiki/ALZ-Policies.md:73]: Update as per policy name and description suggestions

**Review by jtracey93 — CHANGES_REQUESTED** [review]
Thanks @Springstone & @aarunraaj just a few changes please

Thanks

Jack

File: src/resources/Microsoft.Authorization/policyDefinitions/Deny-VMUnmanagedDisk.json (hunk 6 above)
Comment (Springstone) [reply src/resources/Microsoft.Authorization/policyDefinitions/Deny-VMUnmanagedDisk.json:63]: I missed the escaping of the formula. THanks

File: src/resources/Microsoft.Authorization/policyDefinitions/Deny-VMUnmanagedDisk.json (hunk 5 above)
Comment (Springstone) [reply src/resources/Microsoft.Authorization/policyDefinitions/Deny-VMUnmanagedDisk.json]: If we are sure, ok. Added the others back in.

File: src/resources/Microsoft.Authorization/policyDefinitions/Deny-VMUnmanagedDisk.json (hunk 1 above)
Comment (aarunraaj) [reply src/resources/Microsoft.Authorization/policyDefinitions/Deny-VMUnmanagedDisk.json:62]: @jtracey93 , yes this covers the OS disk. The reason behind I selected this is (correct me if I am wrong, so I can update the data disk field), the unmanaged data disks cannot be attached on the VM which has managed disk, so this is out of scope in context. Secondly, when a VM attached an OS disk as unmanaged disk during creation (even if it has a data disk attached), the VM creation will be denied. For an existing VM this virtual machine will be a non-compliant resource. Let me know your thoughts.

Comment (Springstone) [reply src/resources/Microsoft.Authorization/policyDefinitions/Deny-VMUnmanagedDisk.json:62]: @jtracey93 @aarunraaj I tend to agree, I spent an hour trying to deploy an unmanaged disk and it's not easy. Essentially this is not restricted to deploying a VHD from a storage account. My understanding is also that you can not mix managed and unmanaged disk, which makes sense from an availability perspective. Blocking OS deployment should be sufficient I think

Comment (jtracey93) [reply src/resources/Microsoft.Authorization/policyDefinitions/Deny-VMUnmanagedDisk.json:62]: If you are both happy then so am I. Can we please update the names and descriptions to reflect its only OS disks etc. though please. @aarunraaj @Springstone

File: eslzArm/managementGroupTemplates/policyAssignments/DENY-VMUnmanagedDiskPolicyAssignment.json [hunk 8]
```diff
@@ -20,11 +29,11 @@
     "variables": {
         "policyDefinitions": {
             "denyVMUnmanagedDisk": "[concat('/providers/Microsoft.Management/managementGroups/', parameters('topLevelManagementGroupPrefix'), '/providers/Microsoft.Authorization/policyDefinitions/Deny-VM-with-unmanaged-disk')]"
-            },
+        },
         "policyAssignmentNames": {
-            "denyVMUnmanagedDisk": "Deny-VM-with-unmanaged-disk",
-            "description": "Deny virtual machines not using managed disk. It checks the managedDisk property on virtual machine OS, and Datadisk fields.",
-            "displayName": "Deny virtual machines and virtual machine scale sets not using managed disk"
+            "denyVMUnmanagedDisk": "Deny-UnmanagedDsk",
```
Comment (jtracey93) [inline eslzArm/managementGroupTemplates/policyAssignments/DENY-VMUnmanagedDiskPolicyAssignment.json:34]:
```suggestion
            "denyVMUnmanagedDisk": "Deny-UnmanagedDisk",
```

**Review by jtracey93 — CHANGES_REQUESTED** [review]

File: eslzArm/eslzArm.json [hunk 9]
```diff
@@ -809,6 +818,7 @@
             "databricksPip": "[uri(deployment().properties.templateLink.uri, 'managementGroupTemplates/policyAssignments/DENY-DatabricksPipPolicyAssignment.json')]",
             "databricksVnet": "[uri(deployment().properties.templateLink.uri, 'managementGroupTemplates/policyAssignments/DENY-DatabricksVnetPolicyAssignment.json')]",
             "classicResourcesPolicyAssignment": "[uri(deployment().properties.templateLink.uri, 'managementGroupTemplates/policyAssignments/DENY-ClassicResourceTypesPolicyAssignment.json')]",
+            "VMUnmanagedDiskPolicyAssignment": "[uri(deployment().properties.templateLink.uri, 'managementGroupTemplates/policyAssignments/DENY-unmana.json')]",
```
Comment (jtracey93) [inline eslzArm/eslzArm.json:821]:
```suggestion
            "VMUnmanagedDiskPolicyAssignment": "[uri(deployment().properties.templateLink.uri, 'managementGroupTemplates/policyAssignments/DENY-VMUnmanagedDisk.json')]",
```

**Review by jtracey93 — CHANGES_REQUESTED** [review]

File: eslzArm/eslzArm.json [hunk 10]
```diff
@@ -809,6 +818,7 @@
             "databricksPip": "[uri(deployment().properties.templateLink.uri, 'managementGroupTemplates/policyAssignments/DENY-DatabricksPipPolicyAssignment.json')]",
             "databricksVnet": "[uri(deployment().properties.templateLink.uri, 'managementGroupTemplates/policyAssignments/DENY-DatabricksVnetPolicyAssignment.json')]",
             "classicResourcesPolicyAssignment": "[uri(deployment().properties.templateLink.uri, 'managementGroupTemplates/policyAssignments/DENY-ClassicResourceTypesPolicyAssignment.json')]",
+            "VMUnmanagedDiskPolicyAssignment": "[uri(deployment().properties.templateLink.uri, 'managementGroupTemplates/policyAssignments/DENY-VMUnmanagedDisk.json')]",
```
[the remark this answers is not in the document]
Comment (Springstone) [reply eslzArm/eslzArm.json]: DOH *facepalm*

**Comment by jtracey93** [comment]
Tested fine 
image


Cleaning up manually due to #1199

**Pull request merged**